File: Customer_Segmentation.R
rm(list=ls(all=T))

setwd("C:/Users/RISHI MUKUNTHAN/Desktop/Data Science/Projects/Edwisor Customer Segmentation/Final/R")

getwd()

#Load Libraries with help of lapply function
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", 'sampling', 'DataCombine', 'inTrees')

lapply(x, require, character.only = TRUE)
rm(x)
#Load data into data object(data frame)
data = read.csv("credit-card-data.csv", header = T, na.strings = c(" ", "", "NA"))
#Structure of data
str(data)
#Column Names
colnames(data)
#view head
head(data,5)
#View stats of each column
summary(data)

##################################Missing Values Analysis###############################################
#Don't impute if missing percent more than 30
#Create data frame with missing value count in each column
missing_val = data.frame(apply(data,2,function(x){sum(is.na(x))}))
#New column with column name so that we can drop index
missing_val$Columns = row.names(missing_val)
#Renaming count column
names(missing_val)[1] =  "Missing_percentage"
#Convert to percent
missing_val$Missing_percentage = (missing_val$Missing_percentage/nrow(data)) * 100
#Descending order
missing_val = missing_val[order(-missing_val$Missing_percentage),]
#row name index change
row.names(missing_val) = NULL
#interchange column order
missing_val = missing_val[,c(2,1)]
#Save to csv
#write.csv(missing_val, "Missing_perc_R.csv", row.names = F)

data_output = data

data = subset(data, select = -c(CUST_ID))
# kNN Imputation
library(DMwR)
data_imputed = knnImputation(data, k = 3)

sum(is.na(data))
sum(is.na(data_imputed))

data_imputed$MA_PURCHASE = data_imputed$PURCHASES / data_imputed$TENURE
data_imputed$MA_CASH_ADVANCE = data_imputed$CASH_ADVANCE / data_imputed$TENURE
data_imputed$LIMIT_USAGE = data_imputed$BALANCE / data_imputed$CREDIT_LIMIT
data_imputed$PAY_MINPAY_RATIO = data_imputed$PAYMENTS / data_imputed$MINIMUM_PAYMENTS

x = data_imputed$ONEOFF_PURCHASES
y = data_imputed$INSTALLMENTS_PURCHASES
w = ifelse((x == 0) & (y == 0), "NONE", ifelse((x > 0) & (y > 0), "BOTH",
    ifelse((x > 0) & (y == 0), "ONE_OFF", ifelse((x == 0) & (y > 0),"INSTALLMENT",""))))

data_imputed$PURCHASE_TYPE = w
table(data_imputed$PURCHASE_TYPE)


# dummify the data
library(caret)
dmy <- dummyVars(" ~ .", data = data_imputed)
data_encoded <- data.frame(predict(dmy, newdata = data_imputed))

# data with KPIs
data_kpi = subset(data_encoded, select = -c(PURCHASES,CASH_ADVANCE,
                                            TENURE,BALANCE,CREDIT_LIMIT,PURCHASE_TYPEBOTH,
                                            PAYMENTS,MINIMUM_PAYMENTS))

# data with KPIs
data_kpi_original = subset(data_encoded, select = -c(PURCHASES,CASH_ADVANCE,
                                            TENURE,BALANCE,CREDIT_LIMIT,PURCHASE_TYPEBOTH,
                                            PAYMENTS,MINIMUM_PAYMENTS))

rm(dmy,missing_val,w,x,y)

#Log transformation
data_kpi$BALANCE_FREQUENCY=log(data_kpi$BALANCE_FREQUENCY+1)
data_kpi$ONEOFF_PURCHASES=log(data_kpi$ONEOFF_PURCHASES+1)
data_kpi$INSTALLMENTS_PURCHASES=log(data_kpi$INSTALLMENTS_PURCHASES+1)
data_kpi$PURCHASES_FREQUENCY=log(data_kpi$PURCHASES_FREQUENCY+1)
data_kpi$ONEOFF_PURCHASES_FREQUENCY=log(data_kpi$ONEOFF_PURCHASES_FREQUENCY+1)
data_kpi$PURCHASES_INSTALLMENTS_FREQUENCY=log(data_kpi$PURCHASES_INSTALLMENTS_FREQUENCY+1)
data_kpi$CASH_ADVANCE_FREQUENCY=log(data_kpi$CASH_ADVANCE_FREQUENCY+1)
data_kpi$CASH_ADVANCE_TRX=log(data_kpi$CASH_ADVANCE_TRX+1)
data_kpi$PURCHASES_TRX=log(data_kpi$PURCHASES_TRX+1)
data_kpi$PRC_FULL_PAYMENT=log(data_kpi$PRC_FULL_PAYMENT+1)
data_kpi$MA_PURCHASE=log(data_kpi$MA_PURCHASE+1)
data_kpi$MA_CASH_ADVANCE=log(data_kpi$MA_CASH_ADVANCE+1)
data_kpi$LIMIT_USAGE=log(data_kpi$LIMIT_USAGE+1)
data_kpi$PAY_MINPAY_RATIO=log(data_kpi$PAY_MINPAY_RATIO+1)

data_scaled = data_kpi
#Normalization
for(i in colnames(data_kpi)){
  print(i)
  data_scaled[,i] = (data_kpi[,i] - min(data_kpi[,i]))/
    (max(data_kpi[,i] - min(data_kpi[,i])))
}
#PCA
# Applying PCA
# install.packages('caret')

# install.packages('e1071')
library(caret)
library(e1071)
pca = preProcess(x = data_scaled, method = 'pca', pcaComp = 4)

data_pca = predict(pca, data_scaled)

# Fitting K-Means to the dataset
set.seed(0)
kmeans = kmeans(x = data_pca, centers = 4)
res_kmeans = kmeans$cluster

pdf("Customer Segmentation R Plots.pdf")
# Visualising the clusters

library(factoextra)
fviz_cluster(kmeans, data_pca[, 1:2], ellipse.type = "norm")
fviz_cluster(kmeans, data_scaled[, 2:3], ellipse.type = "norm")

data_output$Group = res_kmeans
data_imputed$Group = res_kmeans
data_kpi_original$Group = res_kmeans

library(dplyr)
data_grouped <- group_by(data_kpi_original, Group)
data_summary = summarise(data_grouped, MA_CASH_ADVANCE = mean(MA_CASH_ADVANCE),
          MA_PURCHASE = mean(MA_PURCHASE), PURCHASES_TRX=mean(PURCHASES_TRX),
          LIMIT_USAGE = mean(LIMIT_USAGE), PAY_MINPAY_RATIO = mean(PAY_MINPAY_RATIO),
          PRC_FULL_PAYMENT = mean(PRC_FULL_PAYMENT), ONEOFF_PURCHASES = mean(ONEOFF_PURCHASES),
          INSTALLMENTS_PURCHASES = mean(INSTALLMENTS_PURCHASES))
#BarPlot
Cluster <- data_summary$Group
Cash_Advance <- data_summary$MA_CASH_ADVANCE
Monthly_Purchase <- data_summary$MA_PURCHASE
Limit_Used <- data_summary$LIMIT_USAGE
PurchasePerTransaction <- data_summary$PURCHASES_TRX
Pay_MinPay_R <- data_summary$PAY_MINPAY_RATIO
Full_Due_Paid <- data_summary$PRC_FULL_PAYMENT
Oneoff <- data_summary$ONEOFF_PURCHASES
Installments <- data_summary$INSTALLMENTS_PURCHASES

# Plot the bar chart 
barplot(Cash_Advance,names.arg=Cluster,xlab="Cluster",ylab="Cash_Advance",col="blue",
        main="Cash_Advance",border="red")

barplot(Limit_Used,names.arg=Cluster,xlab="Cluster",ylab="Limit_Used",col="blue",
        main="Limit_Used",border="red")

barplot(Monthly_Purchase,names.arg=Cluster,xlab="Cluster",ylab="Monthly_Purchase",col="blue",
        main="Monthly_Purchase",border="red")

barplot(PurchasePerTransaction,names.arg=Cluster,xlab="Cluster",ylab="PurchasePerTransaction",col="blue",
        main="PurchasePerTransaction",border="red")

barplot(Pay_MinPay_R,names.arg=Cluster,xlab="Cluster",ylab="Pay_MinPay_R",col="blue",
        main="Pay_MinPay_R",border="red")

barplot(Full_Due_Paid,names.arg=Cluster,xlab="Cluster",ylab="Full_Due_Paid",col="blue",
        main="Full_Due_Paid",border="red")

#x11()
Purchase_Types <- rbind(Oneoff,Installments)

barplot(Purchase_Types,beside=T,names = c("C1", "C2","C3","C4"),
        xlab="Cluster",ylab="Purchases",col = c("red", "green"),
        main="Purchase Type",border="red",
        legend = c("Oneoff", "Installments"))

data_output$Group[data_output$Group %in% "1"] = "Withdrawers"
data_output$Group[data_output$Group %in% "2"] = "Installment_Purchasers"
data_output$Group[data_output$Group %in% "3"] = "One_Off_Purchasers"
data_output$Group[data_output$Group %in% "4"] = "Big_Spenders"

# Writing a csv (output)
write.csv(data_output, "Clustering_Output_R.csv", row.names = F)
dev.off()

File: Customer_Segmentation.py
from fancyimpute import KNN
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
#%matplotlib inline
# to display all the columns of the dataframe in the notebook
pd.pandas.set_option('display.max_columns', None)

# load dataset
data = pd.read_csv('credit-card-data.csv')
data.drop('CUST_ID',axis=1,inplace=True)

#Apply KNN imputation algorithm
data = pd.DataFrame(KNN(k = 3).fit_transform(data), columns = data.columns)

#Variables with Missing Value percentage
data.apply(lambda x: sum(x.isnull()/len(data))*100)

data.to_csv('credit_card_knn_imputed.csv', index = False)

data['MA_PURCHASE'] = data['PURCHASES']/data['TENURE']
data['MA_CASH_ADVANCE'] = data['CASH_ADVANCE']/data['TENURE']
data['LIMIT_USAGE'] = data['BALANCE']/data['CREDIT_LIMIT']
data['PAY_MINPAY_RATIO'] = data['PAYMENTS']/data['MINIMUM_PAYMENTS']

#drop purchases,cash_advance,tenure(less variability),Balance,CreditLimit

def purchase_type(data):
    if (data['ONEOFF_PURCHASES']==0) & (data['INSTALLMENTS_PURCHASES']==0):
        return 'NONE'
    if (data['ONEOFF_PURCHASES']>0) & (data['INSTALLMENTS_PURCHASES']>0):
         return 'BOTH'
    if (data['ONEOFF_PURCHASES']>0) & (data['INSTALLMENTS_PURCHASES']==0):
        return 'ONE_OFF'
    if (data['ONEOFF_PURCHASES']==0) & (data['INSTALLMENTS_PURCHASES']>0):
        return 'INSTALLMENT'

data['PURCHASE_TYPE']=data.apply(purchase_type,axis=1)

data = pd.concat([data,pd.get_dummies(data['PURCHASE_TYPE'],drop_first=True)],axis=1)
data_profile = data.copy()
data.drop(columns=['PURCHASE_TYPE','PURCHASES','CASH_ADVANCE','TENURE','BALANCE','CREDIT_LIMIT'
                  ,'PAYMENTS','MINIMUM_PAYMENTS'],axis=1,inplace=True)


# log tranformation
data_logT =data[['BALANCE_FREQUENCY', 'ONEOFF_PURCHASES',
       'INSTALLMENTS_PURCHASES', 'PURCHASES_FREQUENCY',
       'ONEOFF_PURCHASES_FREQUENCY', 'PURCHASES_INSTALLMENTS_FREQUENCY',
       'CASH_ADVANCE_FREQUENCY', 'CASH_ADVANCE_TRX', 'PURCHASES_TRX',
       'PRC_FULL_PAYMENT','MA_PURCHASE', 'MA_CASH_ADVANCE', 'LIMIT_USAGE',
       'PAY_MINPAY_RATIO']].applymap(lambda x: np.log(x+1))

data_logT = pd.concat([data_logT,data[['NONE','ONE_OFF','INSTALLMENT']]],axis=1)

from sklearn.preprocessing import  MinMaxScaler
mm=MinMaxScaler()
data_scaled=mm.fit_transform(data_logT)

data_profile.groupby('PURCHASE_TYPE').apply(lambda x: np.mean(x['MA_PURCHASE'])).plot.barh(color=['m','g','b','y'])
plt.title('Average Monthly Purchase by Purchase Types')
plt.savefig("Monthly_Average_Purchase_by_PurchaseType.png")
plt.show()

data_profile.groupby('PURCHASE_TYPE').apply(lambda x: np.mean(x['MA_CASH_ADVANCE'])).plot.barh(color=['m','g','b','y'])
plt.title('Average Cash Advance amount by Purchase Types')
plt.savefig("Cash_Advance_by_PurchaseType.png")
plt.show()

data_profile.groupby('PURCHASE_TYPE').apply(lambda x: np.mean(x['PAY_MINPAY_RATIO'])).plot.barh(color=['m','g','b','y'])
plt.title('Mean Payment_MinPayment Ratio by Purchase Type')
plt.savefig("Pay_MinPay_Ratio_by_PurchaseType.png")
plt.show()

from sklearn.decomposition import PCA

var_ratio={}
for n in range(2,17):
    pc=PCA(n_components=n)
    data_pca=pc.fit(data_scaled)
    var_ratio[n]=sum(data_pca.explained_variance_ratio_)
    
var_ratio

pd.Series(var_ratio).plot(title='Number of Principal Components vs Explained Variance')
plt.savefig("PCA Explained Variance Graph.png")
plt.show()

pca_final=PCA(n_components=4).fit(data_scaled)
PC_data=pca_final.fit_transform(data_scaled)
PC_data=pd.DataFrame(PC_data)

# Factor Analysis : variance explained by each component- 
pd.Series(pca_final.explained_variance_ratio_,index=['PC_'+ str(i) for i in range(1,5)])

PCA_insight = pd.DataFrame(pca_final.components_.T, columns=['PC_' +str(i) for i in range(1,5)],index=data.columns)
PCA_insight.abs().sort_values('PC_1',ascending=False)

PCA_insight.abs().sort_values('PC_2',ascending=False)

#Load required libraries
from sklearn.cluster import KMeans

#Estimate optimum number of clusters
cluster_range = range( 1, 20 )
cluster_errors = []

for num_clusters in cluster_range:
    clusters = KMeans(num_clusters).fit(PC_data.iloc[:,:])
    cluster_errors.append(clusters.inertia_)
    
#Create dataframe with cluster errors
clusters_df = pd.DataFrame( { "num_clusters":cluster_range, "cluster_errors": cluster_errors } )

#Plot line chart to visualise number of clusters
#%matplotlib inline  
plt.figure(figsize=(12,6))
plt.xlabel("No of Clusters")
plt.ylabel("WCSS")
plt.plot( clusters_df.num_clusters, clusters_df.cluster_errors, marker = "o" )
plt.savefig("No of Clusters vs WCSS.png")
plt.show()

#Implement kmeans
kmeans_model = KMeans(n_clusters = 4,random_state=0).fit(PC_data.iloc[:,:])

idx = np.argsort(kmeans_model.cluster_centers_.sum(axis=1))
lut = np.zeros_like(idx)
lut[idx] = np.arange(4)

kmeans_model.labels_

# Conactenating labels found through Kmeans with data 
data_cluster=pd.concat([data[['MA_PURCHASE','MA_CASH_ADVANCE','LIMIT_USAGE','PAY_MINPAY_RATIO','PURCHASES_TRX'
                             ,'PRC_FULL_PAYMENT','ONEOFF_PURCHASES','INSTALLMENTS_PURCHASES']],
                        pd.Series(lut[kmeans_model.labels_],name='Group')],axis=1)

data_cluster['Group'].unique()

# Mean value gives a good indication of the distribution of data. So we are finding mean value for each variable for each cluster
data_kpi_mean=data_cluster.groupby('Group')\
.apply(lambda x: x[['MA_PURCHASE','MA_CASH_ADVANCE','LIMIT_USAGE','PAY_MINPAY_RATIO',
                   'PURCHASES_TRX','PRC_FULL_PAYMENT','ONEOFF_PURCHASES','INSTALLMENTS_PURCHASES']].mean()).T
data_kpi_mean

color_map={0:'r',1:'b',2:'g',3:'y'}
label_color=[color_map[l] for l in kmeans_model.labels_]
plt.figure(figsize=(7,7))
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.scatter(PC_data.iloc[:,0],PC_data.iloc[:,1],c=label_color,cmap='Spectral',alpha=0.1)
plt.savefig("Clustering Results PC1 vs PC2.png")
plt.show()

color_map={0:'r',1:'b',2:'g',3:'y'}
label_color=[color_map[l] for l in kmeans_model.labels_]
plt.figure(figsize=(7,7))
plt.xlabel('One Off Purchases')
plt.ylabel('Installment Purchases')
plt.scatter(data_scaled[:,1],data_scaled[:,2],c=label_color,cmap='Spectral',alpha=0.1)
plt.savefig("Clustering Results OneOff_Purchases vs Installment Purchases.png")
plt.show()

Counts =data_cluster.groupby('Group').apply(lambda x: x['Group'].value_counts())
print (Counts)

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

MA_CASH_ADVANCE=(data_kpi_mean.loc['MA_CASH_ADVANCE',:].values)

bar_width=.10
b1=plt.bar(index,MA_CASH_ADVANCE,color='r',label='MA_CASH_ADVANCE',width=bar_width)

plt.xlabel("Clusters")
plt.title("Mean MA_Cash_Advance among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis1.png")
plt.show()

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

LIMIT_USAGE=(data_kpi_mean.loc['LIMIT_USAGE',:].values)

bar_width=.10
b1=plt.bar(index,LIMIT_USAGE,color='r',label='LIMIT_USAGE',width=bar_width)

plt.xlabel("Cluster")
plt.title("Mean Limit Usage among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis2.png")
plt.show()

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

MA_PURCHASE=(data_kpi_mean.loc['MA_PURCHASE',:].values)

bar_width=.10
b1=plt.bar(index,MA_PURCHASE,color='b',label='MA_PURCHASE',width=bar_width)

plt.xlabel("Cluster")
plt.title("Mean MA_Purchase among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis3.png")
plt.show()

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

PURCHASES_TRX=(data_kpi_mean.loc['PURCHASES_TRX',:].values)

bar_width=.10
b1=plt.bar(index,PURCHASES_TRX,color='b',label='PURCHASES_TRX',width=bar_width)

plt.xlabel("Clusters")
plt.title("Mean Purchase/Txn among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis4.png")
plt.show()

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

PRC_FULL_PAYMENT=(data_kpi_mean.loc['PRC_FULL_PAYMENT',:].values)

bar_width=.10
b1=plt.bar(index,PRC_FULL_PAYMENT,color='b',label='PRC_FULL_PAYMENT',width=bar_width)

plt.xlabel("Clusters")
plt.title("Full_Payment_Due_Percentage among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis5.png")
plt.show()

fig,ax=plt.subplots(figsize=(5,3))
index=np.arange(len(data_kpi_mean.columns))

PAY_MINPAY_RATIO=(data_kpi_mean.loc['PAY_MINPAY_RATIO',:].values)

bar_width=.10
b1=plt.bar(index,PAY_MINPAY_RATIO,color='b',label='PAY_MINPAY_RATIO',width=bar_width)

plt.xlabel("Clusters")
plt.title("Average Pay_MinPay_Ratio Among Clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis6.png")
plt.show()

fig,ax=plt.subplots(figsize=(8,5))
index=np.arange(len(data_kpi_mean.columns))

ONEOFF_PURCHASES=(data_kpi_mean.loc['ONEOFF_PURCHASES',:].values)
INSTALLMENTS_PURCHASES=(data_kpi_mean.loc['INSTALLMENTS_PURCHASES',:].values)

bar_width=.10
b1=plt.bar(index,ONEOFF_PURCHASES,color='b',label='ONEOFF_PURCHASES',width=bar_width)
b2=plt.bar(index+bar_width,INSTALLMENTS_PURCHASES,color='r',label='INSTALLMENTS_PURCHASES',width=bar_width)

plt.xlabel("Clusters")
plt.title("Purchases - OneOff vs Installment among clusters")
plt.xticks(index, ('Cl-0', 'Cl-1', 'Cl-2', 'Cl-3'))
plt.legend()
plt.savefig("Segmentation Analysis7.png")
plt.show()

data_label = pd.read_csv('credit_card_knn_imputed.csv')
data_raw = pd.read_csv('credit-card-data.csv')
# Conactenating labels found through Kmeans with data 
data_label=pd.concat([data_raw['CUST_ID'],data_label,
                        pd.Series(lut[kmeans_model.labels_],name='Group')],axis=1)

map_cluster = {0: "Withdrawers", 1: "Installment_Purchasers", 2: "One_Off_Purchasers", 3: "Big_Spenders"}
data_label = data_label.replace({"Group": map_cluster})

data_label.to_csv('Clustering_Output_Python.csv', index = False)